### ISSUE TITLE
Contrast Manager Failing During Statistics-Volume Pipeline 

### ISSUE THREAD

**Issue by J-V1** [aramis-lab/clinica#976]
When running the pipeline, it doesn't not complete. Instead it gives the following error: 

![error2](https://github.com/aramis-lab/clinica/assets/129490230/19150a6e-cf15-4b41-949c-98057a8fe5c1)

I am using matlab 2017b and followed the format of the constrat tsv provided in the documentation:

![Screenshot 2023-09-26 105845](https://github.com/aramis-lab/clinica/assets/129490230/4b095ca8-f788-4620-8884-cd7e910d6253)

Note the age used for the subjects are arbitrary, just testing if this works as I am new to clinica. Here is what the command I used looks like: 

clinica run statistics-volume CAPS_DIRECTORY AD pet-volume DATA.TSV age --acq_label 18FAV1451 --suvr_reference_region pons --n_procs 3

any help would be greatly appreciated as i have been stuck on this for weeks now to no avail. thank you <3

**Comment by NicolasGensollen** [comment]
Hi @J-V1 

Thanks for reporting. 
I haven't looked in details, but at first glance I'm a bit confused by the `""` around the age values in the TSV file. 
This might be totally unrelated, but it could be worth trying to rerun the pipeline after having cleaned the age column with proper floats.

**Comment by J-V1** [comment]
Hey @NicolasGensollen thank you for your response. I originally put the quotes around the age to make it read as a string as I get this type error when I don't:

![screenshot 2](https://github.com/aramis-lab/clinica/assets/129490230/c6b781c6-b0de-4bf9-82ad-79082f8aa517)

Please let me know if there is a solution for this or if you need any additional info from me!

**Comment by NicolasGensollen** [comment]
Hi @J-V1 

I had a quick look at your issue. The problem comes from the fact that the statistics volume pipeline expects to have only two groups which will then be compared through the contrast. 
You can clearly see this when looking at this function which is responsible for finding the two class names as well as their indices:

https://github.com/aramis-lab/clinica/blob/13e091a784318deebd77dc9e243df63ba68f1e59/clinica/pipelines/statistics_volume/statistics_volume_utils.py#L18-L37

If you try to use it directly on your TSV file with "age" as the contrast:

```python
>>> from clinica.pipelines.statistics_volume.statistics_volume_utils import get_group_1_and_2
>>> idx1, idx2, classes = get_group_1_and_2("./group-UnitTest_covariates.tsv", "age")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/nicolas.gensollen/GitRepos/clinica-fork/clinica/pipelines/statistics_volume/statistics_volume_utils.py", line 54, in get_group_1_and_2
    raise ClinicaException(
clinica.utils.exceptions.ClinicaException: It must exist only 2 classes in the column age to perform 2-sample t-tests. Here Clinica found: ['55.9', '56.2', '58.2', '59.3', '59.9', '60.6', '63.8', '65.1', '66.5', '66.7', '67.5', '68', '69.2', '69.7', '69.8', '70', '70.4', '70.5', '70.8', '70.9', '71', '71.2', '71.4', '72.1', '72.5', '72.6', '73.1', '73.4', '73.5', '73.6', '74.2', '74.4', '75.1', '75.4', '75.7', '75.8', '76.4', '76.5', '77.2', '78,0', '78.1', '78.2', '78.6', '78.7', '79', '79.4', '79.6', '81', '81.1', '81.7', '81.9', '82.1', '82.3', '83.2', '83.6', '85.8', '86.5']
```

Using "sex" as the contrast:

```python
>>> idx1, idx2, classes = get_group_1_and_2("./group-UnitTest_covariates.tsv", "sex")
>>> classes
['Female', 'Male']
```

That being said, I must admit that the documentation is misleading, especially when it says:

- `CONTRAST` is a string defining the contrast matrix or the variable of interest for the GLM, e.g. group or age.

I will investigate further to verify that I am correct and will rectify the documentation once sure.

If you want to experiment the pipeline, you should be able to proceed by specifying a proper contrast such as "group" or "sex":

```
clinica run statistics-volume CAPS_DIRECTORY AD pet-volume DATA.TSV group --acq_label 18FAV1451 --suvr_reference_region pons --n_procs 3
```

Please let me know if this helps.

**Comment by J-V1** [comment]
Thank you for your response! Sadly even when running the command with "group" and "sex" respectively, it led to the same error message.

![issue persists](https://github.com/aramis-lab/clinica/assets/129490230/303c2362-bbb0-407d-9717-869b6adfe38a)

Do you think this could be an issue outside of clinica, i.e matlab not being set up correctly? Or maybe do you have a tsv that 100% works that I can test on my system? 

Appreciate all the help once again!

**Comment by NicolasGensollen** [comment]
Hi @J-V1 

It is strange that you get the same error.
I'm not sure whether this is a problem in Clinica, SPM/Matlab or the input data at this point. So let's try to find out:

- Could you tell me which version of Clinica and Python you are using ?
- Are you using Matlab + SPM or SPM standalone ?
- Were you able to use Matlab and/or SPM before and without Clinica on your machine ? (I'm a bit concerned by the LibGL error message you have in your logs)
- What does your input data look like ?
- Did you run the `PETVolume` pipeline before trying to run the `StatisticsVolume` pipeline as described in the docs ?

**Comment by J-V1** [comment]
Thank you for your continued help! It should be noted that I am working in a WSL.

-Currently using  clincia v0.75 and python v3.8.17 

-Using both Matlab + SPM

-Able to use matlab and spm out of WSL in clinica enviornment without any issues.

-Input data is just the subjects and sessions from the TSV i attached in the original post. I used the pet-volume and t1-volume and the outputs of those were used for StatisticsVolume. More specifically each of the sessions M-120 for sub-ADNI041S1418 and M-120 for sub-ADNI141S1378 have the atlas_statistics folder as well as all of the expected PET images. I have attached an image of my file structure below if that at all helps. 
![example](https://github.com/aramis-lab/clinica/assets/129490230/b96cdb10-32c1-48cf-b24d-f9a1fdddfca5)

-Able to run PETvolume without any major issues. Screenshot attached below!
![pet volume](https://github.com/aramis-lab/clinica/assets/129490230/942c65c1-13cb-4131-a294-35aabaf057c8)

Error is still persisting even after re-running StatisticsVolume with pons2 data. Also, something I forgot to mention is it says when the pipeline is running that "Images generated by SPM will popup during the pipeline." However, this never occurs. Not sure if this is indictive of anything, but figured it wouldn't hurt to add. 

Very appreciative of you taking the time to look at this issue, fingers crossed we can get to the bottom of it!

**Comment by NicolasGensollen** [comment]
Hi @J-V1 

Thanks for your reply. I must admit that I don't know whether running this pipeline within a WSL will work or not.

What you can try is to run the following script. It basically reproduces the first steps of the statistics pipeline up to the point where a SPM model file is created and run. Note that you will have to change a few paths to match your input data.

Let me know how it goes and what you are able to obtain from this script (standard output as well as potential errors that you'd get, especially when calling `run_m_script()` at the end.

```python
from pathlib import Path
from clinica.pipelines.statistics_volume.statistics_volume_utils import (
    get_group_1_and_2,
    run_m_script,
    write_matlab_model,
)
from clinica.utils.inputs import clinica_file_reader
from clinica.utils.participant import get_subject_session_list
from clinica.utils.pet import Tracer
from clinica.utils.input_files import pet_volume_normalized_suvr_pet

def main():
    # You need to update these to match your data inputs
    # Path to the clinica folder
    clinica_install_dir = Path("/Users/nicolas.gensollen/GitRepos/clinica")

    # Path to where the data is located
    input_dir = Path("/Users/nicolas.gensollen/GitRepos/clinica_data_ci/data_ci/StatisticsVolume/in")
    caps_dir = input_dir / "caps"
    # The group label
    group_label = "UnitTest"
    # You might need to change the name of the TSV file below to match yours
    tsv_file = input_dir / f"group-{group_label}_covariates.tsv"

    contrast = "group"
    template_file = clinica_install_dir / "clinica" / "pipelines" / "statistics_volume" / "template_model_creation.m"

    assert tsv_file.exists()
    assert template_file.exists()

    query = pet_volume_normalized_suvr_pet(
        acq_label= Tracer.FDG,
        group_label="*",
        suvr_reference_region="pons",
        use_brainmasked_image=True,
        use_pvc_data=False,
        fwhm=8,
    )
    sessions, subjects = get_subject_session_list(caps_dir, tsv_file, is_bids_dir=False)
    files, _ = clinica_file_reader(subjects, sessions, caps_dir, query)
    assert len(files) > 0
    print("Pipeline will run on the following files: " + "\n-".join(files))

    idx_group1, idx_group2, class_names = get_group_1_and_2(tsv_file, contrast)
    print(f"Classes for t-test are {class_names}")

    model_file, covariates = write_matlab_model(tsv_file, contrast, idx_group1, idx_group2, files, template_file)
    print(f"{model_file=}")
    assert Path(model_file).exists()
    
    run_m_script(model_file)


if __name__ == "__main__":
    main()
```

**Comment by J-V1** [comment]
I was able to run the script with some small adaptations, but got this error stating that the output matrix was not produced.

![Output matrix not produced](https://github.com/aramis-lab/clinica/assets/129490230/af7f514b-6d6c-4884-920c-dbb368be9ad8)

I decided to further examine this and it appears that SPM is unable to identify the  NIfTI images for the two groups. I double checked the paths and permissions and they line up to my knowledge, but I still continue to get this errorr.

![no matching files](https://github.com/aramis-lab/clinica/assets/129490230/2cc7c7c3-a951-4a4b-9a18-b1286b423d02)

I am going to continue trying some different set ups and moving my files off from windows drive and into the subsystem itself to see if that helps. Thank you for all your help once again!
